db/db_impl/db_impl.cc

```diff
@@ -149,6 +149,24 @@ void DumpSupportInfo(Logger* logger) {
 
   ROCKS_LOG_HEADER(logger, "DMutex implementation: %s", DMutex::kName());
 }
+
+// `start` is the inclusive lower user key bound without user-defined timestamp
+// `limit` is the exclusive upper user key bound without user-defined timestamp
+std::tuple<Slice, Slice> MaybeAddTimestampsToRange(const Slice& start,
+                                                   const Slice& limit,
+                                                   size_t ts_sz,
+                                                   std::string* start_with_ts,
+                                                   std::string* limit_with_ts) {
+  if (ts_sz == 0) {
+    return std::make_tuple(start, limit);
+  }
+  // Maximum timestamp means including all key with any timestamp
+  AppendKeyWithMaxTimestamp(start_with_ts, start, ts_sz);
+  // Append a maximum timestamp as the range limit is exclusive:
+  // [start, limit)
+  AppendKeyWithMaxTimestamp(limit_with_ts, limit, ts_sz);
```

Comment:
Just to double check: is it OK to append the max timestamp here because it's really a "don't care" value (since stuff would already get excluded based on the user-provided key part)?

Reply:
This is a great question! I looked a bit into it, unlike the point look up and iterator logic, the approximate size logic for both sst files[1] and memtables[2] are treating user key as a whole, not differentiating  UDT or non-UDT cases. So max timestamp is a safer one to use for this.

[1]:https://github.com/facebook/rocksdb/blob/main/db/version_set.cc#L6732-L6733
[2]:https://github.com/facebook/rocksdb/blob/main/memtable/skiplistrep.cc#L95